src/vsr/client_replies.zig

```diff
@@ -228,33 +241,37 @@ pub fn ClientRepliesType(comptime Storage: type) type {
             callback: fn (client_replies: *ClientReplies) void,
         ) void {
             assert(client_replies.ready_callback == null);
-            client_replies.ready_callback = callback;
 
             if (client_replies.writes.available() > 0) {
+                client_replies.ready_callback = .{ .next_tick = callback };
                 client_replies.storage.on_next_tick(
                     ready_next_tick_callback,
                     &client_replies.ready_next_tick,
                 );
             } else {
                 // ready_callback will be called the next time a write completes.
+                client_replies.ready_callback = .{ .write = callback };
             }
         }
 
         fn ready_next_tick_callback(next_tick: *Storage.NextTick) void {
             const client_replies = @fieldParentPtr(ClientReplies, "ready_next_tick", next_tick);
+            assert(client_replies.writes.available() > 0);
 
-            if (client_replies.ready_callback) |callback| {
-                assert(client_replies.writes.available() > 0);
+            const callback = client_replies.ready_callback.?.next_tick;
+            client_replies.ready_callback = null;
+            callback(client_replies);
+        }
 
-                client_replies.ready_callback = null;
-                callback(client_replies);
-            } else {
-                // Another write finished before the next_tick triggered.
-            }
+        pub fn remove_reply(client_replies: *ClientReplies, slot: Slot) void {
```

Comment:
Hm, when we call this from `replica`, what guarantees that we don't have any in-flight ops for the Slot? We could assert that here?

Reply:
I think it doesn't actually matter if we do have in-flight writes to this op.
Writes will queue up anyway, and `faulty` is only set (by the replica) after checking `ClientSessions`, which is always up-to-date.

Reply:
In the read path, when we fetch message from in-flight write, we assert that the checksum matches. Can we hir that?

Reply:
I'm not sure what you mean :thinking: 
Are you suggesting we pass the entry into `remove_reply`, and assert that the latest write (if it exists) matches that entry?
(Ignoring `header.size == sizeOf(Header)` I guess.)

Reply:
Oh just realized that s/hir/hit

But now I don't understand how we could hit that assert :sweat_smile:

Reply:
Yeah, I think I convinced myself that it is indeed unreachable